Aliasing of build.rs output when `--out-dir` is used

<!-- Thanks for filing a 🐛 bug report 😄! -->

**Problem**

Unstable flag `--out-dir` is unsafe in presence of multiple packages with build scripts & parallel build.

What I think is happening is with flag `--out-dir` cargo places compiled `build.rs` in a same location for all packages. For example, if I set `--out-dir=./target`, all build scripts will be compiled into `target/build-script-build`.

In case of a parallel build, this causes race condition between multiple processes. Eventually, build fails, either because script is deleted by another process or when wrong script is executed.

When build scripts do things like setting extra rustc arguments or generate files, build fails in a particularly spectacular way.

As an example, this is what we periodically get on our CI:

```
error[E0277]: the trait bound `i128: traits::Num` is not satisfied
   --> /usr/local/cargo/registry/src/github.com-1ecc6299db9ec823/num-integer-0.1.39/src/roots.rs:136:14
    |
136 |         impl Roots for $T {
    |              ^^^^^ the trait `traits::Num` is not implemented for `i128`
...
170 | signed_roots!(i128, u128);
    | -------------------------- in this macro invocation
```
I think, this happens because num-traits uses build.rs to set `has_i128`. So, if "wrong" build.rs is being executed, package does not get this flag and eventually its dependency `num-integer` fails because of an error above (because it sets this flag independently).

**Steps**

1. Get sample project from https://github.com/idubrov/cargo-out-dir-bug
2. This project is nothing more than a 100 empty modules with `build.rs` files.
3. Run `cargo build` -- everything works fine
4. Run `cargo -Z unstable-options build --out-dir=./target`
5. Eventually, build fails with:

```
error: failed to remove file `/Users/idubrov/Projects/outdir/./target/build-script-build`                                                                                                                                                     

Caused by:
  No such file or directory (os error 2)
```

**Possible Solution(s)**

Maybe, treat `build.rs` output specially from regular outputs? Never place it into the directory specified by `--out-dir`?

**Notes**

Output of `cargo version`:

```
$ cargo version
cargo 1.31.0-nightly (2d0863f65 2018-10-20)
```

--- a/src/cargo/core/compiler/mod.rs
+++ b/src/cargo/core/compiler/mod.rs
@@ -437,11 +437,13 @@ fn link_targets<'a, 'cfg>(
             destinations.push(dst.display().to_string());
             hardlink_or_copy(src, dst)?;
             if let Some(ref path) = export_dir {
-                if !path.exists() {
-                    fs::create_dir_all(path)?;
-                }
+                if !target.is_custom_build() {
+                    if !path.exists() {
+                        fs::create_dir_all(path)?;
+                    }
 
-                hardlink_or_copy(src, &path.join(dst.file_name().unwrap()))?;
+                    hardlink_or_copy(src, &path.join(dst.file_name().unwrap()))?;
+                }
             }
         }
 

--- a/tests/testsuite/out_dir.rs
+++ b/tests/testsuite/out_dir.rs
@@ -203,6 +203,37 @@ fn replaces_artifacts() {
     .run();
 }
 
+#[test]
+fn avoid_build_scripts() {
+    let p = project()
+        .file(
+            "Cargo.toml",
+            r#"
+            [workspace]
+            members = ["a", "b"]
+            "#,
+        )
+        .file("a/Cargo.toml", &basic_manifest("a", "0.0.1"))
+        .file("a/src/main.rs", "fn main() {}")
+        .file("a/build.rs", r#"fn main() { println!("hello-build-a"); }"#)
+        .file("b/Cargo.toml", &basic_manifest("b", "0.0.1"))
+        .file("b/src/main.rs", "fn main() {}")
+        .file("b/build.rs", r#"fn main() { println!("hello-build-b"); }"#)
+        .build();
+
+    p.cargo("build -Z unstable-options --out-dir out -vv")
+        .masquerade_as_nightly_cargo()
+        .with_stdout_contains("[a 0.0.1] hello-build-a")
+        .with_stdout_contains("[b 0.0.1] hello-build-b")
+        .run();
+    check_dir_contents(
+        &p.root().join("out"),
+        &["a", "b"],
+        &["a", "a.dSYM", "b", "b.dSYM"],
+        &["a.exe", "a.pdb", "b.exe", "b.pdb"],
+    );
+}
+
 fn check_dir_contents(
     out_dir: &Path,
     expected_linux: &[&str],

